specification/sections/01-flag-evaluation.md

```diff
@@ -171,6 +171,16 @@ client.getMetadata().getDomain(); // "domain-1"
 In previous drafts, this property was called `name`.
 For backwards compatibility, implementations should consider `name` an alias to `domain`.
 
+#### Requirement 1.2.3
+
+> The client interface **MUST** define a `allFlagMetadata` member or accessor, returning a collection of flag metadata from the provider.
+
+```typescript
+client.getAllFlagMetadata() // [{"key": "featureA", "type": "boolean"}, {"key": "featureB", "type": "string"}]
+```
+
+This may return an empty collection if the provider does not implement the `allFlagMetadata` member or accessor.
```

Comment:
I am not sure about this one @maxveldink as I would assume something like an optional, null, undefined, or error code if the provider does not implement it.
Having an empty collections would mean that semantically, "no flags available" is the same as "flag metadata" not implemented. 
As a consumer this could be problematic because one might have to implement different behavior if `allFlagsMetadata` is not implemented compared to just no flags being available.

Reply:
One of the rules we have is that none of the client methods should throw - the client is the primary abstraction the `application author` interfaces with - it's the most likely abstraction to be used in request handlers and other critical paths at runtime.

I think we want to allow implementations a lot of flexibility here, but I would suggest a null or undefined return that's clearly annotated.

Reply:
> I think we want to allow implementations a lot of flexibility here, but I would suggest a null or undefined return that's clearly annotated.

Yes thats exactly what I was suggesting @toddbaert. Maybe for the wording it could be something like `a language idiomatic way of describing the absence of a value` to kept the implementation open null, undefined, options, maybes, etc.

Reply:
Not throwing makes sense to me. Creating the distinction between a `null` or `undefined` value being returned vs a collection would be enough information for the caller to decide what to do. Unlike flag evaluation, since not all providers will implement this behavior, callers cannot have a hard requirement on the return value (or should always handle the sad path, regardless of the underlying issue).

I've updated the language here to match @lukas-reining 's suggestion. Let me know how that looks!